examples/generative/corrdiff/train.py

```diff
@@ -14,338 +14,336 @@
-    if dist.rank == 0:
-        os.makedirs(c.run_dir, exist_ok=True)
-        with open(os.path.join(c.run_dir, "training_options.json"), "wt") as f:
-            json.dump(c, f, indent=2)
-
     (
         dataset,
-        dataset_iter,
-        valid_dataset,
-        valid_dataset_iter,
+        dataset_iterator,
+        validation_dataset,
+        validation_dataset_iterator,
     ) = init_train_valid_datasets_from_config(
         dataset_cfg,
         data_loader_kwargs,
-        batch_size=batch_size_gpu,
-        seed=seed,
+        batch_size=cfg.training.hp.batch_size_per_gpu,
+        seed=0,
         validation_dataset_cfg=validation_dataset_cfg,
     )
 
-    (img_shape_y, img_shape_x) = dataset.image_shape()
-    if (c.patch_shape_x is None) or (c.patch_shape_x > img_shape_x):
-        c.patch_shape_x = img_shape_x
-    if (c.patch_shape_y is None) or (c.patch_shape_y > img_shape_y):
-        c.patch_shape_y = img_shape_y
-    if c.patch_shape_x != img_shape_x or c.patch_shape_y != img_shape_y:
-        if c.patch_shape_x != c.patch_shape_y:
-            raise NotImplementedError("Rectangular patch not supported yet")
-        if c.patch_shape_x % 32 != 0 or c.patch_shape_y % 32 != 0:
-            raise ValueError("Patch shape needs to be a multiple of 32")
+    # Parse image configuration & update model args
+    dataset_channels = len(dataset.input_channels())
+    img_in_channels = (
+        dataset_channels + cfg.model.args.N_grid_channels
+    )  # noise + low-res input
+    img_shape = dataset.image_shape()
+    img_out_channels = len(dataset.output_channels())
+    if cfg.model.hr_mean_conditioning:
+        img_in_channels += img_out_channels
+
+    # Parse the patch shape
+    if hasattr(cfg, "training.hp.patch_shape_x"):
+        patch_shape_x = cfg.training.hp.patch_shape_x
+    else:
+        patch_shape_x = None
+    if hasattr(cfg, "training.hp.patch_shape_y"):
+        patch_shape_y = cfg.training.hp.patch_shape_y
+    else:
+        patch_shape_y = None
+    patch_shape = (patch_shape_y, patch_shape_x)
+    patch_shape, img_shape = set_patch_shape(img_shape, patch_shape)
+    if patch_shape != img_shape:
         logger0.info("Patch-based training enabled")
     else:
         logger0.info("Patch-based training disabled")
+    # interpolate global channel if patch-based model is used
+    if img_shape[1] != patch_shape[1]:
+        img_in_channels += dataset_channels
+
+    # Parse model args, append additional required args from dataset configs.
+    # Then instantiate the model and move to device.
+    model_args = parse_model_args(cfg.model.args)
+    additional_model_args = {
+        "img_in_channels": img_in_channels,
+        "img_out_channels": img_out_channels,
+        "img_resolution": list(img_shape),
+        "use_fp16": fp16,
+    }
+    model_args.update(additional_model_args)
+    model = Module.instantiate(
+    {
+        "__name__": cfg.model.name,
+        "__args__": model_args,
+    }
```

Comment:
I personally do not like the idea of separating the recipes for different trainings (regression, diffusion, patch-based diffusion) because there will be significant code duplication, and code maintenance would be harder as well. This is the standard way of instantiating the models in Modulus. There is adequate documentation about this, and other examples are using the same approach, including the unified recipe for forecasting models. So consistency between examples is another reason for why I prefer to keep instantiation this way.

Reply:
I agree with Simon FWIW. It is not easy to jump to the model implementation, and I would argue the config is too general since it allows using any modulus model when only a single one (SongUnet) has been tested for this application. I don't think we have plans (or would suggest) that users use other architectures/preconditioners.

It's a common sentiment that people prefer to write python in python rather than yaml.

Maybe we should define what is a "recipe", "example", etc. Generality might be a goal for the "unified recipe", but would argue that clarity is more important for corrdiff.

> This is the standard way of instantiating the models in Modulus 

Do you have feedback from users that they prefer this to clearer (yet less flexible) example scripts?

Reply:
Yes, I agree it might be a good idea to separate "recipes" from "examples". But in that case, I think CorrDiff needs to be a recipe. We have a lot of interest in training custom CorrDiff versions and experimenting with different options. I have upcoming additions myself that can make use of a highly configurable recipe.

I also agree that avoiding code replication should be a priority for maintainability. I'm not personally much bothered by indirection, I find avoiding clutter and DRY / single-point-of-definition principles more important for following code. That said, in this particular case the code is not very obvious and we should make it clear with comments what happens here (i.e. we initialize an instance of class with the given name). Also, perhaps one way to make this more obvious would be to wrap this in a function with reasonable default arguments?

Reply:
Thanks for your comments. For simplicity, I changed the code to explicitly instantiate the models with hard-coded arguments. Please take a look